File: day12/day12.cpp
#include <iostream>
#include <fstream>
#include <vector>

struct Node {
    std::vector<Node*> connections = {};
    std::string name;
    bool big; // Big: true, Small: false
    bool visited = false;
};

int findPaths(std::vector<Node*> nodes, Node* start, Node* end) {
    // Find every path from start to end, passing through big false nodes once and type true nodes as many times as needed
    int paths = 0;
    start->visited = true;
    std::vector<Node*> visited = {start};
    std::vector<Node*> toVisit = start->connections;
    while (toVisit.size() > 0) {
        // print out to visit nodes
        for (auto node : toVisit) {
            std::cout << node->name << " ";
        }
        std::cout << std::endl;
        Node* current = toVisit.back();
        toVisit.pop_back();
        if (current == end) {
            paths++;
        }
        current->visited = true;
        for (Node* n : current->connections) {
            if (current->big || !current->visited) {
                toVisit.push_back(n);
                current->visited = true;
            }
        }
    }
    return paths;
}

int main()
{
    // Setup
    std::ifstream file("input0.txt");
    std::string line;
    std::vector<Node*> nodes;
    while (std::getline(file, line)) {
        Node* n = new Node;
        Node* e = new Node;
        int split = line.find('-');
        std::string nodeName = line.substr(0, split);
        std::string edgeName = line.substr(split+1, line.size()-2);

        bool nodeExists = false;
        bool edgeExists = false;
        int nodeIndex = 0;
        int edgeIndex = 0;
        for (int i = 0; i < nodes.size(); i++) {
            if (nodes[i]->name == nodeName) {
                nodeExists = true;
                nodeIndex = i;
            }
            if (nodes[i]->name == edgeName) {
                edgeExists = true;
                edgeIndex = i;
            }
        }

        if (nodeExists) {
            n = nodes[nodeIndex];
            // std::cout << "(" << nodeIndex << ") Node exists: " << nodeName << std::endl;
        } else {
            n->name = nodeName;
            n->big = std::isupper(nodeName[0]) ? true : false;
            nodes.push_back(n);
        }

        if (edgeExists) {
            e = nodes[edgeIndex];
            // std::cout << "(" << edgeIndex << ") Edge exists: " << edgeName << std::endl;
        } else {
            e->name = edgeName;
            e->big = std::isupper(nodeName[0]) ? true : false;
            nodes.push_back(e);
        }
        n->connections.push_back(e);
        e->connections.push_back(n);
    }

    for (int i = 0; i < nodes.size(); i++) {
        std::cout << nodes[i]->name << ": (";
        std::cout << nodes[i]->big << ") -> ";
        std::cout << "{ ";
        for (int j = 0; j < nodes[i]->connections.size(); j++) {
            std::cout << nodes[i]->connections[j]->name << "-";
            std::cout << &nodes[i]->connections[j] << ", ";
        }
        std::cout << " }" << std::endl;
    }

    // Part 1
    Node* start;
    Node* end;
    for (int i = 0; i < nodes.size(); i++) {
        if (nodes[i]->name == "start") {
            start = nodes[i];
        }
        if (nodes[i]->name == "end") {
            end = nodes[i];
        }
    }

    int paths = findPaths(nodes, start, end);
    std::cout << paths << std::endl;

    return 0;
}

File: day12/day12.py
import sys

def parseInput(filename):
    graph = {}
    with open(filename, 'r') as f:
        for line in f.readlines():
            line = line.strip()
            vecs = line.split('-')
            if vecs[0] in graph:
                graph[vecs[0]].append(vecs[1])
            else:
                graph[vecs[0]] = [vecs[1]]
            if vecs[1] in graph:
                graph[vecs[1]].append(vecs[0])
            else:
                graph[vecs[1]] = [vecs[0]]
    return graph

def count_paths(graph: dict, current: str, visited: set):
    if (current == "end"): return 1
    
    paths = 0
    for vec in graph[current]:
        if vec not in visited:
            if not vec.isupper(): visited.add(vec)
            paths += count_paths(graph, vec, visited)
            if not vec.isupper(): visited.remove(vec)
    
    return paths

def count_paths2(graph: dict, current: str, visited: set, double_small: bool):
    if (current == "end"): return 1
    
    paths = 0
    for vec in graph[current]:
        if vec not in visited:
            if not vec.isupper(): visited.add(vec)
            paths += count_paths2(graph, vec, visited, double_small)
            if not vec.isupper(): visited.remove(vec)
        elif not double_small and vec != "start":
            paths += count_paths2(graph, vec, visited, True)
    
    return paths

def main(part):
    graph = parseInput('input.txt')
    
    if part == 1 or part == -1:
        paths = count_paths(graph, "start", {"start"})
        print(f"Part 1:\nThere are {paths} paths.")

    if part == 2 or part == -1:
        paths = count_paths2(graph, "start", {"start"}, False)
        print(f"Part 2:\nThere are {paths} paths.")
    


if __name__ == "__main__":
    part = -1

    if (len(sys.argv) > 1):
        part = int(sys.argv[1])
    
    if part not in {-1, 1, 2}:
        print("Usage: day12.py [part]")
        print("\tpart: 1 or 2")
        exit(1)
    else:
        main(part)